### ISSUE TITLE
Overflows can occur with generated quantities

### ISSUE THREAD

**Issue by mortonjt** [gibsramen/BIRDMAn#45]
Right now the way that many of these stan programs are built is that the generated quantities block will return a ton of nans.

For instance if you run [negative_binomial_single.stan](https://github.com/gibsramen/BIRDMAn/blob/main/birdman/templates/negative_binomial_single.stan#L37), you won't be able to run any of the arviz diagnostics due to the nans in `y_predict`.

These NaNs stem from the underlying rate Poisson rate parameter, apparently if the rate parameter is ever greater than 21 (i.e. `log(2^31)`), you are going to hit [integer overflow](https://mc-stan.org/docs/2_21/reference-manual/numerical-data-types.html).  See this [post](https://discourse.mc-stan.org/t/multivariate-poisson-model-having-mysterious-nans-cryptic-errors/1490).  

I see 3 different solutions
1. Have the user fine-tune the priors to make sure that the rate parameter never gets too high.  
2. Force the rate parameter to be below 20
3. Forget about sampling from negative binomial, use a smooth approximation so that we can return floats instead of ints.

Not sure what the best way to proceed is. I guess the real question is why are we getting such high rate parameters?

**Comment by gibsramen** [comment]
This concords with my experience as well. For one, I think we could add a check and subsequent warnings if there are NaNs when converting to InferenceData as this has burned me a few times.

In terms of solutions, I'm not really sure. I've just been fine-tuning priors but if there were a more robust solution or we could figure out why the rate parameters are so high that would be ideal.

**Comment by mortonjt** [comment]
I think I have a solution posted here : https://github.com/flatironinstitute/q2-matchmaker/issues/24

The issue here (I think) is that the intercept priors are way off (for the single nb regression models, ALR regression is a different story) - turns out that if your Poisson rate parameter is greater than 20, you will nana (since exp(20) is the integer overflow limit).   
The intercepts are ultimately modeling the log-proportions of your reference group, and log proportions are strictly less than 0 and shouldn't be centered at zero.  My benchmarks are suggesting that mu=-5 and std=3 is is a much more appropriate prior for the intercept, and I don't have any more nans.

**Comment by gibsramen** [comment]
Gotcha - that certainly makes sense. Given that, do you think we could have a "default" prior calculated as a function of the table? Or do you think N(-5, sd=3) is generally appropriate across datasets/omics?

**Comment by mortonjt** [comment]
Definitely not appropriate across datasets / omics. I was thinking that we could hard-code these intercept priors depending on the omics dataset (i.e. amplicon, shotgun, metabolomics, RNAseq...). 

The other approach we could do is to provide a empirical bayes helper function, which would essentially wrap the code that I wrote in https://github.com/flatironinstitute/q2-matchmaker/issues/24 to grab the mean and standard deviations from the log proportions of the observed counts.  There may need to be some tweaking required since this won't for zeros, etc.

**Comment by mortonjt** [comment]
Here's an update -- apparently those priors are actually quite consistent across datasets, I'm getting similar results between amplicon, shotgun and RNAseq.

**Comment by mortonjt** [comment]
I'm going to close this, since it is more of a problem with the stan file.

There are two possible solutions
1. Set a reasonable prior on log proportions
2. Apply a log softmax or log sigmoid to strictly ensure that the proportions are correctly bounded.
